### PR TITLE
refactor(connector): new profile system, decoupling the integration and profile module, import the injection framework

### PR DESCRIPTION
**Pull request by TalkWIthKeyboard**
For more information, see the [issue](https://github.com/tellery/tellery/issues/318).

### CODE REVIEW

**Review by raycursive — APPROVED** [review]

**Comment by raycursive** [comment]
ref #132

File: packages/connector/interface/src/integrations/DbtIntegration.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+)
+abstract class DbtIntegration : BaseIntegration() {
+
+    companion object {
+        val mapper = jacksonObjectMapper()
+    }
+
+    protected fun getValueOrThrowException(profile: NewProfile, key: String): String {
+        return profile.configs[key]
+            ?: throw RuntimeException("Required key $key is not in ${profile.type} dataset profile.")
```
Comment (TalkWIthKeyboard) [inline packages/connector/interface/src/integrations/DbtIntegration.kt:44]: use a custom exception

File: packages/connector/interface/src/integrations/DbtIntegration.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    open fun transformToDbtProfile(profile: NewProfile): BaseDbtProfile {
+        throw NotImplementedError("The dbt integration is not impl transformToDbtProfile method.")
+    }
+}
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+open class BaseDbtProfile(open val type: String)
+
+annotation class Type(val value: String)
```
Comment (TalkWIthKeyboard) [inline packages/connector/interface/src/integrations/DbtIntegration.kt:55]: Type -> DbtIntegrationType

File: packages/connector/server/build.gradle.kts [hunk 3]
```diff
@@ -69,9 +72,17 @@ dependencies {
...
-    testImplementation(kotlin("test"))
+    implementation("org.koin:koin-core:$koinVersion")
+    implementation("org.ktorm:ktorm-core:$ktormVersion")
+    implementation("org.ktorm:ktorm-support-postgresql:$ktormVersion")
+
+    testImplementation("io.kotest:kotest-runner-junit5:$kotestVersion")
+    testImplementation("io.kotest:kotest-assertions-core:$kotestVersion")
+    testImplementation("io.kotest:kotest-property:$kotestVersion")
+    testImplementation("io.kotest:kotest-framework-datatest:$kotestVersion")
+    testImplementation("io.kotest.extensions:kotest-extensions-testcontainers:1.0.1")
```
Comment (TalkWIthKeyboard) [inline packages/connector/server/build.gradle.kts:83]: add version to project file.

File: packages/connector/server/src/Main.kt [hunk 4]
```diff
@@ -68,6 +70,11 @@ class ConnectorServer(
 }
 
 fun main() {
+    startKoin {
```
Comment (TalkWIthKeyboard) [inline packages/connector/server/src/Main.kt:73]: remove test code

File: packages/connector/server/src/Modules.kt [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import io.tellery.managers.impl.FileProfileManager
+import io.tellery.services.DbtV2Service
+import io.tellery.services.ProfileService
+import io.tellery.services.RpcService
+import org.koin.core.module.Module
+import org.koin.core.qualifier.named
+import org.koin.dsl.module
+import entities.ProjectConfig as config
+
+object Modules {
```
Comment (TalkWIthKeyboard) [inline packages/connector/server/src/Modules.kt:17]: merge the code to `MainV2`

File: packages/connector/interface/src/annotations/Integration.kt [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package annotations
```
Comment (raycursive) [inline packages/connector/interface/src/annotations/Integration.kt:1]:
```suggestion
package io.tellery.annotations
```

Comment (raycursive) [inline packages/connector/interface/src/annotations/Integration.kt:1]:
```suggestion
package io.tellery.annotations
```

File: packages/connector/interface/src/entities/ProjectConfig.kt [hunk 7]
```diff
@@ -0,0 +1,54 @@
+package entities
```
Comment (raycursive) [inline packages/connector/interface/src/entities/ProjectConfig.kt:1]:
```suggestion
package io.tellery.entities
```
I'm guessing that the best practice in the kotlin style guide might not be followed by the autocomplete of idea.
maybe we should fall back to package-based folder structure ?

File: packages/connector/interface/src/entities/Resource.kt [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package entities
```
Comment (raycursive) [inline packages/connector/interface/src/entities/Resource.kt]:
```suggestion
package io.tellery.entities
```

File: packages/connector/interface/src/entities/Resource.kt [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+data class NewProfile(
+    // unique id
+    var id: String,
+    var type: String,
+    var credential: String?,
+    var configs: Map<String, String?>
+)
+
+data class Integration(
```
Comment (raycursive) [inline packages/connector/interface/src/entities/Resource.kt:11]: will this naming be conflict with the annotation?

File: packages/connector/implementation/src/integrations/dbt/BigQueryDbtIntegration.kt [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+            val keyfileJson: BigQueryConnector.BigQueryKeyBody = mapper.readValue(jsonBody)
+
+            return BigQueryDbtProfile(
+                method = "service-account-json",
+                priority = "interactive",
+                dataset = getValueOrThrowException(it, BigQueryFields.DATASET),
+                keyfileJson = keyfileJson,
+                project = keyfileJson.projectId
+            )
+        }
```
Comment (raycursive) [inline packages/connector/implementation/src/integrations/dbt/BigQueryDbtIntegration.kt:29]:
```suggestion
        val jsonBody = Base64
            .getDecoder()
            .decode(getValueOrThrowException(profile, BigQueryFields.KEY_FILE))
            .decodeToString()

        val keyfileJson: BigQueryConnector.BigQueryKeyBody = mapper.readValue(jsonBody)

        return BigQueryDbtProfile(
            method = "service-account-json",
            priority = "interactive",
            dataset = getValueOrThrowException(it, BigQueryFields.DATASET),
            keyfileJson = keyfileJson,
            project = keyfileJson.projectId
        )
```
meaningless let here?

File: packages/connector/server/src/managers/DbtManagerV2.kt [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+        .associateBy { it.findAnnotation<Type>()!!.value }
+
+    /**
+     * Connector and context are a read more write fewer resources, so we store the state in memory
+     * to replace requesting from the database every time.
+     * Dbt profile file is a resource that needs to be async with the latest profile, and it is base
+     * on the file system, so we need to get a write lock before update it.
+     *
+     * In summary, we require a read-write lock in this class, get write lock when reload context and
+     * get read lock when do other dbt operate.
```
Comment (raycursive) [inline packages/connector/server/src/managers/DbtManagerV2.kt:63]:
```suggestion
     * Connector and context are read-more-write-fewer resources, so we store the state in memory
     * to replace requesting from the database each time.
     * Dbt profile file is a resource that needs to be async with the latest profile, and it is based
     * on the file system, so we need to get a write lock before updating it.
     *
     * In summary, we require a read-write lock in this class, get write lock when reloading context and
     * get read lock when doing other dbt operations.
```

File: packages/connector/server/src/managers/DbtManagerV2.kt [hunk 12]
```diff
@@ -0,0 +1,397 @@
...
+            return
+        }
+
+        connector = profile
+        context = Context(globalRepoDir, keyDir, integration)
+
+        /**
+         * FileUtils.forceMkdir() ignores the situation that dir is exists;
+         * File.createNewFile() returns false while the file is exists;
+         * so it needn't to check the file is exists.
```
Comment (raycursive) [inline packages/connector/server/src/managers/DbtManagerV2.kt:93]:
```suggestion
         * FileUtils.forceMkdir() ignores the situation that dir exists;
         * File.createNewFile() returns false while the file exists;
         * so it doesn't need to check if the file exists.
```

File: packages/connector/server/src/managers/DbtManagerV2.kt [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+        val sources = manifest.sources.values.filter {
+            it.config.enabled && it.resourceType == "source"
+        }.map { it.toDbtBlock() }
+
+        return models + sources
+    }
+
+    /**
+     * Insure the "dbt_project.yml" file has the valid "model" node,
+     * if it need update the project file, the function will push the diff to remote repository after updating file.
```
Comment (raycursive) [inline packages/connector/server/src/managers/DbtManagerV2.kt:213]:
```suggestion
     * if it needs to update the project file, the function will push the diff to the remote repository after updating the file.
```

**Review by raycursive — APPROVED** [review]

File: packages/connector/interface/src/entities/ProjectConfig.kt (hunk 7 above)
Comment (TalkWIthKeyboard) [reply packages/connector/interface/src/entities/ProjectConfig.kt:1]: sure, it is more convenient.

File: packages/connector/implementation/src/integrations/dbt/BigQueryDbtIntegration.kt (hunk 10 above)
Comment (TalkWIthKeyboard) [reply packages/connector/implementation/src/integrations/dbt/BigQueryDbtIntegration.kt:29]: It was used in two places.

File: packages/connector/server/src/entities/ProjectConfig.kt [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    val port: Int
+        get() = env.getOrDefault("server.port", "50051").toInt()
+
+    val credCertPath: String?
+        get() = env["server.credential.certificate"]
+
+    val credKeyPath: String?
+        get() = env["server.credential.key"]
+
+    val deployModel: DeployModel
```
Comment (raycursive) [inline packages/connector/server/src/entities/ProjectConfig.kt:21]: deployMode maybe  better?

**Pull request merged**